### DISCUSSION TITLE
Timers without Ticking

### DISCUSSION THREAD

**Discussion by smessmer** [bevyengine/bevy#7593] [Ideas]
Both the `Timer` and `Stopwatch` APIs need you to tick them, i.e. periodically update them in one of your systems. Is this actually necessary?

Imagine a `Timer` class that - instead of storing a `elapsed: Duration` member just stores a `start: Instant` member. Such a timer wouldn't need to be ticked, but we'd have to change the `elapsed()` function to take a `Time&` argument so that we can calculate it on the fly. Because of that, it's not a clear win but it's a different way of doing this.

In my personal opinion, however, that's a better way because it is more explicit. You cannot accidentally forget to tick the timer or accidentally tick it twice and get weird behavior in your game. Also, it would likely be more accurate since we don't have as many floating point addition operations.

**Comment by akimakinai** [comment]
That said, real-time independent and explicitly ticked timers might work better when pausing the game, changing its speed, or saving, loading or rolling back the game state. Also, since duration is represented by `(u64, u32)`, ticking by duration seems to be integer addition.

**Reply by smessmer** [reply]
You're right that the implementation details of the timer get a bit more complicated to handle pausing and speed changes but it's possible and the API should look the same.

**Reply by smessmer** [reply]
Oh actually the API would have to change. We'd have to pass in a `&Time` object to `pause()` and `unpause()` so `pause` can remember when it was paused and `unpause` can adjust the `start` member accordingly.

**Comment by smessmer** [comment]
I guess another option is to have timers automatically register themselves with a central timer registry and make that registry tick them automatically. But that comes with some overhead for creating timers.
